### PR TITLE
Modified ProjectileLoader.CanDamage call.

### PR DESCRIPTION
**Pull request by ThomasSpeedrunner**
<s>CanDamage if set to true now skips the massive fuckfest of an individual projectile ID blacklist at the start of Damage()
added showCrit, showSpeed, and showMana to simplify making certain weapons show or not show crit chance, use speed, and mana cost</s>

CanDamage is now a nullable wrapper (thank you to Ozzatron for teaching me this terminology)
setting it to true allows a projectile to do damage no matter what, setting it to false disallows a projectile from doing damage no matter what, and leaving it unset lets vanilla call the shots

### CODE REVIEW

File: patches/tModLoader/Terraria/Projectile.cs.patch [hunk 1]
```diff
@@ -112,19 +112,24 @@
...
  				return;
-+			
+ 
++			SkipWhateverTheHellThisHardcodedGroceryListIs:
 +			if (Main.projPet[type] && !ProjectileLoader.MinionContactDamage(this))
 +				return;
 +			
-+			if (!ProjectileLoader.CanDamage(this))
-+				return;
-
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/Projectile.cs.patch:125]: You do realize that you just removed vanilla logic in its entirety?

**Review by Mirsario — COMMENTED** [review]
Use PR templates and save memeing for discord.

File: patches/tModLoader/Terraria/Projectile.cs.patch (hunk 1 above)
Comment (Mirsario) [reply patches/tModLoader/Terraria/Projectile.cs.patch:125]: And you made it actually do the damage when it returns false.

**Review by Mirsario — CHANGES_REQUESTED** [review]
Marking my previous comments as change requests. That should be rolled back. We can talk about tooltips though.

File: patches/tModLoader/Terraria/Projectile.cs.patch [hunk 2]
```diff
@@ -343,6 +342,27 @@
...
...pe == 600 || type == 601 || type == 602 || type == 535 || (type == 631 && localAI[1] == 0f) || type == 651 || (type == 188 && localAI[0] < 5f) || (aiStyle == 137 && ai[0] != 0f) || aiStyle == 138 || (type == 261 && velocity.Length() < 1.5f) || (type == 818 && ai[0] < 1f) || type == 831 || (type == 833 && ai[0] == 4f) || (type == 834 && ai[0] == 4f) || (type == 835 && ai[0] == 4f) || (type == 281 && ai[0] == -3f) || ((type == 598 || type == 636 || type == 614) && ai[0] == 1f) || (type == 923 && localAI[0] <= 60f) || (type == 919 && localAI[0] <= 60f) || (aiStyle == 15 && ai[0] == 0f && localAI[1] <= 12f) || type == 861 || (aiStyle == 93 && ai[0] != 0f && ai[0] != 2f) || (aiStyle == 10 && localAI[1] == -1f) || (Main.projPet[type] && type != 266 && type != 407 && type != 317 && (type != 388 || ai[0] != 2f) && (type < 390 || type > 392) && (type < 393 || type > 395) && (type != 533 || !(ai[0] >= 6f) || !(ai[0] <= 8f)) && (type < 625 || type > 628) && (type != 755 || ai[0] == 0f) && (type != 946 || ai[0] == 0f) && type != 758 && type != 951 && (type != 759 || frame == Main.projFrames[type] - 1) && type != 833 && type != 834 && type != 835 && type != 864 && (type != 623 || ai[0] != 2f)))
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Projectile.cs.patch:354]: This kind of patch is to be avoided if possible (which it is). I'm actually not quite sure why we needed a separate method here, couldn't the vanilla code be kept, and just move the ProjectlieLoader.CanDamage call above it?

Comment (Mirsario) [reply patches/tModLoader/Terraria/Projectile.cs.patch:354]: Yes, this is super unnecessary and unmaintainable. Don't move or rewrite vanilla code, work around it.

**Review by Mirsario — CHANGES_REQUESTED** [review]
I had a small discussion in discord, and everyone there at the time seemed to agree with the following - I think there's a way better implementation for tooltip toggling to be done. After tooltips receive somewhat similar treatment to player layers (1.4_playerlayers), we can most likely force all vanilla tooltips be registered at all times, but be hidden based on original logic. That way, you'll be able to enable whatever tooltips you would like by just finding them in the ModifyTooltips' provided dictionary under "Terraria" index, and make them visible by just setting a boolean to true. So, remove the tooltip patches. I'll merge the bool? CanDamage change, it seems good now.

**Comment by ThomasSpeedrunner** [comment]
should be good now

File: patches/tModLoader/Terraria/Main.cs.patch [hunk 3]
```diff
@@ -1523,6 +1524,15 @@
...
 +						toolTipLine[numLines] = crit + Lang.tip[5].Value;
++						toolTipNames[numLines] = "CritChance";
++						numLines++;
++					}
++					*/
++					if (!item.summon) {
++						int crit = player[myPlayer].GetWeaponCrit(item);
++						ItemLoader.GetWeaponCrit(item, player[myPlayer], ref crit);
++						PlayerHooks.GetWeaponCrit(player[myPlayer], item, ref crit);
++						toolTipLine[numLines] = crit + Lang.tip[5].Value;
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/Main.cs.patch:1535]: You need to revert this change too. I'm also happy to merge when you do. This is an example of why it's best to split features into separate PRs because one might be approved and the other need to wait for a redesign after discussion

Comment (ThomasSpeedrunner) [reply patches/tModLoader/Terraria/Main.cs.patch]: fixed up, should be good for real this time now

**Comment by ThomasSpeedrunner** [comment]
blep. pretty sure it's fine now

File: patches/tModLoader/Terraria/ModLoader/GlobalProjectile.cs [hunk 4]
```diff
@@ -158,12 +158,15 @@ public class GlobalProjectile:ModType
 		}
 
 		/// <summary>
-		/// Whether or not the given projectile is capable of killing tiles (such as grass) and damaging NPCs/players. Return false to prevent it from doing any sort of damage. Returns true by default.
+		/// Whether or not the given projectile is capable of killing tiles (such as grass) and damaging NPCs/players.
+		/// Return false to prevent it from doing any sort of damage.
+		/// Return true if you want the projectile to do damage regardless of the default blacklist.
+		/// Return nothing to let the projectile follow vanilla can-damage-anything rules. This is what happens by default.
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/GlobalProjectile.cs:164]:
```suggestion
		/// Return null to let the projectile follow vanilla can-damage-anything rules. This is what happens by default.
```

File: patches/tModLoader/Terraria/ModLoader/ModProjectile.cs [hunk 5]
```diff
@@ -234,10 +234,13 @@ public class ModProjectile : ModTexturedType
 		}
 
 		/// <summary>
-		/// Whether or not this projectile is capable of killing tiles (such as grass) and damaging NPCs/players. Return false to prevent it from doing any sort of damage. Returns true by default.
+		/// Whether or not this projectile is capable of killing tiles (such as grass) and damaging NPCs/players.
+		/// Return false to prevent it from doing any sort of damage.
+		/// Return true if you want the projectile to do damage regardless of the default blacklist.
+		/// Return nothing to let the projectile follow vanilla can-damage-anything rules. This is what happens by default.
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ModProjectile.cs:240]:
```suggestion
		/// Return null to let the projectile follow vanilla can-damage-anything rules. This is what happens by default.
```

File: patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs [hunk 6]
```diff
@@ -295,18 +295,17 @@ private class HookList
...
 			foreach (GlobalProjectile g in HookCanDamage.arr) {
-				if (!g.Instance(projectile).CanDamage(projectile)) {
-					return false;
+				bool? canDamage = g.Instance(projectile).CanDamage(projectile);
+				if (canDamage.HasValue) {
+					return canDamage.Value;
 				}
 			}
-			return true;
+			return projectile.modProjectile?.CanDamage();
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs:308]: As this uses nullable bools, this method should probably not short-circuit so simply. I think it should prefer returning false, if any hooks returned it? Then true, then null.

File: patches/tModLoader/Terraria/Projectile.cs.patch [hunk 7]
```diff
@@ -146,19 +146,27 @@
...
  					projectile.projUUID = projectile.identity;
  
-@@ -8831,6 +_,12 @@
+@@ -8829,9 +_,20 @@
+ 		}
+ 
  		public void Damage() {
++			bool? shouldHitAnything = ProjectileLoader.CanDamage(this);
++			if (shouldHitAnything.HasValue) {
++				if (shouldHitAnything.Value == true)
```
Comment (Mirsario) [inline patches/tModLoader/Terraria/Projectile.cs.patch:155]: herbert.png
```suggestion
+				if (shouldHitAnything.Value)
```

**Review by Mirsario — CHANGES_REQUESTED** [review]

**Review by Mirsario — APPROVED** [review]
@Chicken-Bones what do you think about the current ProjectileLoader.CanDamage?

File: patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs [hunk 8]
```diff
@@ -295,18 +295,24 @@ private class HookList
...
-					return false;
+				bool? canDamage = g.Instance(projectile).CanDamage(projectile);
+
+				if (canDamage.HasValue) {
+					if (!canDamage.Value) {
+						return false;
+					}
+
+					result &= canDamage.Value;
 				}
```
Comment (Chicken-Bones) [inline patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs:312]: This doesn't work, because the starting value is null. This should just be = true

**Review by Chicken-Bones — CHANGES_REQUESTED** [review]
Good PR. Please fix last thing then merge.

Also @Mirsario I think we could've done some of the final cleanup for him, rather than having a lot of small back and forth. Poor guy's got a lot of commits here

File: patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs (hunk 8 above)
Comment (ThomasSpeedrunner) [reply patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs:312]: what even is &= as an operator :thinkies:

Comment (Mirsario) [reply patches/tModLoader/Terraria/ModLoader/ProjectileLoader.cs:312]: a = a & b.
